python/amici/sbml_import.py

```diff
@@ -1473,6 +1482,105 @@ def _parse_special_functions(sym: sp.Expr, toplevel: bool = True) -> sp.Expr:
+
+    :param args:
+        Arguments to the piecewise function
+
+    :return:
+        Arguments where conditions no longer contain piecewise functions and
+        the coditional dependency is flattend out
+    """
+    args_out = []
+    for coeff, cond in grouper(args, 2, True):
+        # handling of this case is explicitely disabled in
+        # _parse_special_functions as keeping track of coeff/cond
+        # arguments is tricky. Simpler to just parse them out here
+        if coeff.__class__.__name__ == 'piecewise':
+            coeff = _parse_special_functions(coeff, False)
+
+        # we can have conditions that are piecewise function
+        # returning True or False
+        if cond.__class__.__name__ == 'piecewise':
+            # this keeps track of conditional that the previous
+            # piece was picked
+            previous_was_picked = sp.false
+            # recursively denest those first
+            for sub_coeff, sub_cond in grouper(
+                    _denest_piecewise(cond.args), 2, True
+            ):
+                # flatten the individual pieces
+                pick_this = sp.And(
+                    sp.Not(previous_was_picked), sub_cond
+                )
+                if sub_coeff == sp.true:
+                    args_out.extend([coeff, pick_this])
+                previous_was_picked = pick_this
+
+        else:
+            args_out.extend([coeff, cond])
+    # cut off last condition as that's the default
+    return tuple(args_out[:-1])
+
+def _parse_piecewise_to_heaviside(args: Iterable[sp.Expr]) -> sp.Expr:
+    """
+   Piecewise functions cannot be transformed into C++ right away, but AMICI has
+   a special interface for Heaviside function, so we transform them
+
+    :param args:
+        symbolic expressions for arguments of the piecewise function
+    """
+    # how many condition-expression pairs will we have?
+    formula = sp.Float(0.0)
+    not_condition = sp.Float(1.0)
+
+    def _parse_trigger(trigger: sp.Expr) -> sp.Expr:
+        """
+        Recursively translates a boolean trigger function into a real valued
+        root function
+
+        :param trigger:
+        :return: real valued root function expression
+        """
+        if trigger.is_Relational:
+            root = trigger.args[0] - trigger.args[1]
+            if isinstance(trigger, (sp.core.relational.StrictLessThan,
+                                    sp.core.relational.LessThan)):
+                root *= -1
+        elif isinstance(trigger, sp.Or):
+            return sp.Max(*[_parse_trigger(arg)
+                            for arg in trigger.args])
+        elif isinstance(trigger, sp.And):
+            return sp.Min(*[_parse_trigger(arg)
+                            for arg in trigger.args])
+        else:
+            raise SBMLException('AMICI can not parse piecewise functions '
+                                f'with argument {trigger}.')
+        return root
```

Comment:
Seems overly complex. Simplify?

```suggestion
            return root
        
        if isinstance(trigger, sp.Or):
            return sp.Max(*[_parse_trigger(arg)
                            for arg in trigger.args])
        
        if isinstance(trigger, sp.And):
            return sp.Min(*[_parse_trigger(arg)
                            for arg in trigger.args])
        
        raise SBMLException('AMICI can not parse piecewise functions '
                                f'with argument {trigger}.')
        
```